### PR TITLE
http conn pool: rework closeConnections() -> drainConnections()

### PR DESCRIPTION
**Pull request by mattklein123**
This is a follow up from https://github.com/envoyproxy/envoy/pull/2005.
From production usage at Lyft it has become clear that immediately
closing connections on health check failure is too aggressive. During
normal upstream draining when using the x-envoy-immediate-health-check-fail
header, it is possible to close connections while they still have legitimate
draining requests.

This change reworks the functionality to drain but not close existing
connections. This will make it so that *new* streams occur on new connections,
but existing streams are allowed to finish on their existing connections.
This is the behavior we actually wanted in the first place.

*Risk Level*: Low 
*Testing*: UT
*Docs Changes*: N/A
*Release Notes*: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
@alyssawilk PTAL

File: source/common/http/http2/conn_pool.cc [hunk 1]
```diff
@@ -19,22 +19,20 @@ ConnPoolImpl::ConnPoolImpl(Event::Dispatcher& dispatcher, Upstream::HostConstSha
...
 
   if (draining_client_) {
     draining_client_->client_->close();
   }
+
+  // Make sure all clients are destroyed before we are destroyed.
+  dispatcher_.clearDeferredDeleteList();
 }
 
+void ConnPoolImpl::ConnPoolImpl::drainConnections() { movePrimaryClientToDraining(); }
```
Comment (alyssawilk) [inline source/common/http/http2/conn_pool.cc]: My one concern with this is if you have an H2 server which is in lame duck, wouldn't the proper server behavior to state that it's unhealthy and send goaways?  If so will this increase the odds of movePrimaryClientToDraining multiple times and closing draining_client_?    If so we should probably do a follow-up or open an issue to handle multiple goaways better or we'll still end up killing all active streams.

Comment (htuch) [reply source/common/http/http2/conn_pool.cc]: Don't we only `movePrimaryCilentToDraining()` when `primary_client_` is still set, i.e. https://github.com/envoyproxy/envoy/blob/master/source/common/http/http2/conn_pool.cc#L179, so this would be safe?

Comment (alyssawilk) [reply source/common/http/http2/conn_pool.cc]: I'm not familiar with this section of code yet so I may be off base.  Basically I'm thinking if an endpoint goes unhealthy it should immediately send a GOAWAY (resulting in the primary client draining).  If there's then a new stream before the next health check, we'd create the new connection and either get an immediate GOAWAY (and stomping the draining client with all the active streams) or the health check would result in stomping the draining client, which is the one with all the streams.  Either way I have a concern about the long term goal of serving outstanding requests not being well served with only this PR.   It's a step in the right direction (for servers which nack health but don't send GOAWAYs) hence the LGTM, but if my understanding is correct we should probably have an open issue for continued improvement.

Comment (mattklein123) [reply source/common/http/http2/conn_pool.cc]: Yeah this is a little bit complicated. Technically, we should maintain a list of draining connections (potentially up to some max). Way back in the day I opted for simpler implementation to just have primary/draining. I think (I could be wrong) that in general this is safe for most common cases. The history here goes something like:
1) The original implementation just did/primary draining. When an upstream server was lame ducked/drained, Envoy sends GOAWAY. This was causing GOAWAY thrashing as you point out in the window where upstream HC had not detected failure yet. (Since GOAWAY does not lead to HC fail because we use it during hot restart drains).
2) Because of (1), I added x-envoy-immediate-health-check-fail. This fixed the churning by immediately indicating HC fail in the context of a response. However, it was too aggressive and closed all connections when it was received (effectively).
3) This fix improves on (2), by allowing immediate lame duck / HC fail, but allows for draining the connections.

There are cases in which this might still drop requests. E.g., if there was already a connection in draining, but that is extremely unlikely and almost definitely within allowed error bounds.

The TL/DR is that I *think* we are not going to drop pending responses in any normal case with this change. I'm totally open to follow ups but we should identify an actual sequence that we are concerned about and look at how we would fix it.

Comment (alyssawilk) [reply source/common/http/http2/conn_pool.cc]: ahhhh, so x-envoy-immediate-health-check-fail is the bit I was missing.  I still think there's a decent chance of a race unless I'm missing something.

Let's say we get the GOAWAY (immediate on lameduck) and we we drain the connection. 
 Because the connection isn't flagged as unhealthy (no new response headers yet) it still has good odds of a new stream being assigned.  If we get a new incoming stream on the connection before an outstanding request generates a response (before x-envoy-immediate-health-check-fail) we'll create a new connection.  Then when we get the health check failure on the next response, we'll close the original connection (which has most of the open streams) to drain the new connection (with one open request).    Wouldn't that still leave the flake factor fairly high?  

Again if this brings us from "always close the connection" to "close the connection 50% of the time" it's still worth submitting, but if a race is still likely we should file a PR or document it as an outstanding issue.

Comment (mattklein123) [reply source/common/http/http2/conn_pool.cc]: In the general case in which an upstream sends GOAWAY outside of the context of a response, you are correct, there is still high flake factor. In an Envoy mesh however, Envoy never sends GOAWAY outside of response context. Envoy determines "drain close" right before encoding the headers of a response, so the GOAWAY would be received very likely after (or for sure immediately before) a response carrying x-envoy-immediate-health-check-fail. There is still a very small race condition here due to cross thread posting and some host set computations, etc., but it's very small. How small, I'm not sure. We will know when we deploy and I can update. I will open a new issue to track this conversation.

In general though, I'm not really sure what we could do better here beyond potentially switching to a list of draining connections with some cap?

Comment (alyssawilk) [reply source/common/http/http2/conn_pool.cc]: SGTM.  Plus it'd work better for when upstream isn't Envoy so isn't sending envoy health check headers.
